sim/midas/src/main/scala/midas/passes/fame/FAMETransform.scala

```diff
@@ -147,34 +147,63 @@ object FAMEModuleTransformer {
+      clockLowEnable: Expression,
+      clockHighEnable: Expression,
+      clockBuffer: SignalInfo)
+
     // Multi-clock management step 4: Generate clock buffers for all target clocks
-    val targetClockBufs: Seq[SignalInfo] = clockChannel.ports.map { en =>
+    val clockMetadata: Seq[TargetClockMetadata] = clockChannel.ports.map { en =>
       val enableReg = hostFlagReg(s"${en.name}_enabled", resetVal = UIntLiteral(1))
       val buf = WDefInstance(ns.newName(s"${en.name}_buffer"), DefineAbstractClockGate.blackbox.name)
       val clockFlag = DoPrim(PrimOps.AsUInt, Seq(clockChannel.replacePortRef(WRef(en))), Nil, BoolType)
       val connects = Block(Seq(
         Connect(NoInfo, WRef(enableReg), Mux(WRef(finishing), clockFlag, WRef(enableReg), BoolType)),
         Connect(NoInfo, WSubField(WRef(buf), "I"), WRef(hostClock)),
-        Connect(NoInfo, WSubField(WRef(buf), "CE"), And(And(WRef(enableReg), WRef(finishing)), nReset))))
-      SignalInfo(Block(Seq(enableReg, buf)), connects, WSubField(WRef(buf), "O", ClockType, SourceFlow))
+        Connect(NoInfo, WSubField(WRef(buf), "CE"),
+          Seq(WRef(enableReg), WRef(finishing), nReset, WRef(doneInit)).reduce(And.apply))))
+      TargetClockMetadata(
+        en,
+        WRef(enableReg),
+        clockChannel.replacePortRef(WRef(en)),
+        SignalInfo(Block(Seq(enableReg, buf)), connects, WSubField(WRef(buf), "O", ClockType, SourceFlow))
+      )
     }
 
+    val targetClockBufs = clockMetadata.map(_.clockBuffer)
     // Multi-clock management step 5: Generate target clock substitution map
     def asWE(p: Port) = WrappedExpression.we(WRef(p))
     val replaceClocksMap = (clockChannel.ports.map(p => asWE(p)) zip targetClockBufs.map(_.ref)).toMap
+    val clockLowMap = clockMetadata.map(c => WRef(c.targetSourcePort) -> c.clockLowEnable).toMap
+    val clockHighMap = clockMetadata.map(c => WRef(c.targetSourcePort) -> c.clockHighEnable).toMap
 
     // LI-BDN transformation step 1: Build channels
     // TODO: get rid of the analysis calls; we just need connectivity & annotations
     val portDeps = analysis.connectivity(m.name)
 
-    def genMetadata(info: (String, (Option[Port], Seq[Port]))) = info match {
+    def genMetadata(isInput: Boolean)(info: (String, (Option[Port], Seq[Port]))) = info match {
       case (cName, (Some(clock), ports)) =>
         // must be driven by one clock input port
         // TODO: this should not include muxes in connectivity!
         val srcClockPorts = portDeps.getEdges(clock.name).map(portsByName(_))
         assert(srcClockPorts.size == 1)
         val clockRef = WRef(srcClockPorts.head)
-        val clockFlag = DoPrim(PrimOps.AsUInt, Seq(clockChannel.replacePortRef(clockRef)), Nil, BoolType)
+        val clockFlag = if (isInput) {
+          DoPrim(PrimOps.AsUInt, Seq(clockHighMap(clockRef)), Nil, BoolType)
+        } else {
+          DoPrim(PrimOps.AsUInt, Seq(clockLowMap(clockRef)), Nil, BoolType)
+        }
```

Comment:
Yeah, I really don't understand what `clockHigh` and `clockLow` are and why input channels are High and output channels are Low.  Does the FAME transformed model sample inputs on `posedge` and latch new outputs on `negedge`?

I apologize for starting to learn about the transform guts and asking stupid newb questions in your excellent PR that I'm excited to be able to use.

Reply:
No problem. The names derive from my somewhat poor understanding of how the old Chisel CPP backend worked, they should be renamed. Nothing happens on a clock negedge.

The resulting FAME 1-transformed is essentially a two cycle pipeline (or three depending on how you slice it), that acts on consecutive simulation time steps. The enables define when the a channel will be allowed to handshake again in the next host cycle (i.e., they reset their FiredReg). Essentially the FiredReg is the state for a single-bit FSM on each channel. 

In the first stage, clock tokens are unpacked and fanned out to all input channels according to their clock domain. When a cycle moves from stage 1 -> stage 2, firedRegs for all input channels in clock domains marked to fire in that clock token (essentially a timestep) are reset. 

In the second stage, the clock token is again fanned out, this time to all output channels. (The fired registers for output channels form the second set of pipeline registers). When a cycle moves from stage 2 -> stage 3, a target clock edge for that domain is launched, and output channel FiredRegs are reset.

In stage 3, combinational paths for the freshly fired clock-domain resolve. Setting their FiredRegs as handshakes occur. This will take multiple host cycles if a combinational path spans multiple models. 

That's a brain dump. Hopefully it makes some sense. :|

"Finishing" doesn't really imply a target cycle is finishing anymore so much as it means that handshakes on all output channels and all input channels are happening or have happened in the current host cycle. See `isFiredOrFiring`. In practice, since there is a infinite stream of clock tokens, the only source of stalls are the late arrival of input tokens and backpressure from output channels. Combinational paths manifest as both.  If we think of the following timesteps occupying the pipeline:

Stage1  |  Stage2 | Stage3
Time3   |  Time2   | Time1

Time1 generates a stall if:
1) An output channel is backpressuring (<channel_name)_source_ready is low) 
2) An output channel is combinationally dependent on an input channel that has not yet provided it's token (<CCDep channel name>_sink_valid is low). 
Time2 generates a stall if:
- Not all input tokens for domains scheduled to fire at Time 2 are available. Once the edge fires, these tokens should no longer be observable.

Reply:
In Chisel2-generated-C++-emulator parlance, "clock high" essentially means the phase where posedge-triggered events happen, and "clock low" means triggering the evaluation of the continuous assignments of combinational logic. The reliance on straight-line, topologically sorted code meant each was a very large basic block, so somehow these names were assigned to those blocks (way, way) back in the day.

Anyhow, I don't think they map onto what's being described here very well.